bake: panic value is unknown

Made a typo in my definition and it panics:

```hcl
variable "REPO_SLUG" {
  default = "foo/bar"
}

target "default" {
  cache-from = ["${REPO}:master"]
}
```

```
$ docker buildx bake --print
#1 [internal] load local bake definitions
panic: value is unknown

goroutine 1 [running]:
github.com/zclconf/go-cty/cty.Value.AsString({{{0x2c99178?, 0xc000012861?}}, {0x2430a80?, 0xc000355990?}})
        github.com/zclconf/go-cty@v1.16.0/cty/value_ops.go:1393 +0xf8
github.com/docker/buildx/util/buildflags.isEmpty(...)
        github.com/docker/buildx/util/buildflags/utils.go:53
github.com/docker/buildx/util/buildflags.(*CacheOptions).fromCtyValue(0xc00062c450, {{{0x2c989f0?, 0xc00062c3d8?}}, {0x22c96c0?, 0xc00062c3f0?}}, {0x0, 0x0, 0x0})
        github.com/docker/buildx/util/buildflags/cache_cty.go:29 +0x1b5
github.com/docker/buildx/util/buildflags.(*CacheOptions).FromCtyValue(0x41f5bb0?, {{{0x2c989f0?, 0xc00062c3d8?}}, {0x22c96c0?, 0xc00062c3f0?}}, {0x0?, 0x10?, 0x10?})
        github.com/docker/buildx/util/buildflags/cache_cty.go:17 +0x59
github.com/docker/buildx/bake/hclparser.capsuleValueCapsuleType.func1.1({{{0x2c989f0?, 0xc00062c3d8?}}, {0x22c96c0?, 0xc00062c3f0?}}, {0x0, 0x0, 0x0})
        github.com/docker/buildx/bake/hclparser/type_implied_ext.go:66 +0x111
github.com/zclconf/go-cty/cty/convert.conversionToCapsule.func1({{{0x2c989f0?, 0xc00062c3d8?}}, {0x22c96c0?, 0xc00062c3f0?}}, {0x0?, 0xc0004f3f80?, 0x10?})
        github.com/zclconf/go-cty@v1.16.0/cty/convert/conversion_capsule.go:14 +0x4a
github.com/zclconf/go-cty/cty/convert.getConversion.func1({{{0x2c989f0?, 0xc00062c3d8?}}, {0x22c96c0?, 0xc00062c3f0?}}, {0x0, 0x0, 0x0})
        github.com/zclconf/go-cty@v1.16.0/cty/convert/conversion.go:57 +0x2ae
github.com/zclconf/go-cty/cty/convert.GetConversionUnsafe.retConversion.func1({{{0x2c989f0?, 0xc00062c3d8?}}, {0x22c96c0?, 0xc00062c3f0?}})
        github.com/zclconf/go-cty@v1.16.0/cty/convert/conversion.go:203 +0x34
github.com/zclconf/go-cty/cty/convert.Convert({{{0x2c989f0?, 0xc00062c3d8?}}, {0x22c96c0?, 0xc00062c3f0?}}, {{0x2c98980?, 0xc00062b140?}})
        github.com/zclconf/go-cty@v1.16.0/cty/convert/public.go:51 +0x1fb
github.com/docker/buildx/bake/hclparser/gohcl.DecodeOptions.DecodeExpression({0x2a07b48?, 0x2a08bc8?}, {0x7fc9041bed18, 0xc0004f4820}, 0x2cca330?, {0x25f10c0, 0xc0006053f0})
        github.com/docker/buildx/bake/hclparser/gohcl/decode.go:295 +0x114
github.com/docker/buildx/bake/hclparser/gohcl.DecodeOptions.decodeBodyToStruct({0x2a07b48?, 0x2a08bc8?}, {0x2c98910, 0xc0004f3740}, 0xc00062c018, {0x28598e0?, 0xc000605340?, 0xc00062c018?})
        github.com/docker/buildx/bake/hclparser/gohcl/decode.go:138 +0xda5
github.com/docker/buildx/bake/hclparser/gohcl.DecodeOptions.decodeBodyToValue({0x2a07b48?, 0x2a08bc8?}, {0x2c98910, 0xc0004f3740}, 0xc00062c018, {0x28598e0?, 0xc000605340?, 0x508a8d?})
        github.com/docker/buildx/bake/hclparser/gohcl/decode.go:57 +0xee
github.com/docker/buildx/bake/hclparser/gohcl.DecodeOptions.DecodeBody({0x2a07b48?, 0x0?}, {0x2c98910, 0xc0004f3740}, 0xc00062c018, {0x25efec0?, 0xc000605340?})
        github.com/docker/buildx/bake/hclparser/gohcl/decode.go:30 +0x126
github.com/docker/buildx/bake/hclparser.decodeBody(...)
        github.com/docker/buildx/bake/hclparser/hclparser.go:993
github.com/docker/buildx/bake/hclparser.(*parser).resolveBlock(0xc0004d1880, 0xc000619ba0, 0x0)
        github.com/docker/buildx/bake/hclparser/hclparser.go:456 +0xb85
github.com/docker/buildx/bake/hclparser.Parse({0x2c988d8, 0xc000011968}, {0x2a07b40?, 0xc0004a3680?, 0x2a07b38?}, {0x2657f60, 0xc0004a36b0})
        github.com/docker/buildx/bake/hclparser/hclparser.go:814 +0x1ce7
github.com/docker/buildx/bake.ParseFiles({0xc0004a3650, 0x1, 0x46d31d?}, 0xc0004a3680)
        github.com/docker/buildx/bake/bake.go:347 +0xa5d
github.com/docker/buildx/bake.ReadTargets({0xc0004d1480?, 0x2c983b0?}, {0xc0004a3650?, 0x0?, 0x0?}, {0xc0003552e0, 0x1, 0x0?}, {0x0, 0x0, ...}, ...)
        github.com/docker/buildx/bake/bake.go:197 +0x46
github.com/docker/buildx/commands.runBake({_, _}, {_, _}, {_, _, _}, {{0x426f800, 0x0, 0x0}, ...}, ...)
        github.com/docker/buildx/commands/bake.go:216 +0x185e
github.com/docker/buildx/commands.bakeCmd.func1(0xc0005c0c08, {0xc0003b1350, 0x0, 0x1})
        github.com/docker/buildx/commands/bake.go:454 +0x2e5
github.com/docker/cli/cli-plugins/plugin.RunPlugin.func1.1.2(0xc0005c0c08, {0xc0003b1350, 0x0, 0x1})
        github.com/docker/cli@v27.5.0+incompatible/cli-plugins/plugin/plugin.go:64 +0x6c
github.com/spf13/cobra.(*Command).execute(0xc0005c0c08, {0xc00001bc10, 0x1, 0x1})
        github.com/spf13/cobra@v1.8.1/command.go:985 +0xaaa
github.com/spf13/cobra.(*Command).ExecuteC(0xc000457b08)
        github.com/spf13/cobra@v1.8.1/command.go:1117 +0x3ff
github.com/spf13/cobra.(*Command).Execute(...)
        github.com/spf13/cobra@v1.8.1/command.go:1041
github.com/docker/cli/cli-plugins/plugin.RunPlugin(0xc0002eb900, 0xc0005c0608, {{0x28a30e5, 0x5}, {0x28ad4f4, 0xb}, {0x2c6e770, 0x1a}, {0x0, 0x0}, ...})
        github.com/docker/cli@v27.5.0+incompatible/cli-plugins/plugin/plugin.go:79 +0x145
main.runPlugin(0xc0002eb900)
        github.com/docker/buildx/cmd/buildx/main.go:65 +0xe6
main.run(0xc0002eb900)
        github.com/docker/buildx/cmd/buildx/main.go:79 +0xe5
main.main()
        github.com/docker/buildx/cmd/buildx/main.go:89 +0x86
```

--- a/bake/hcl_test.go
+++ b/bake/hcl_test.go
@@ -2,8 +2,10 @@ package bake
 
 import (
 	"reflect"
+	"regexp"
 	"testing"
 
+	hcl "github.com/hashicorp/hcl/v2"
 	"github.com/stretchr/testify/require"
 )
 
@@ -647,7 +649,7 @@ func TestHCLAttrsCapsuleType(t *testing.T) {
 	require.Equal(t, []string{"default", "key=path/to/key"}, stringify(c.Targets[0].SSH))
 }
 
-func TestHCLAttrsCapsuleTypeVars(t *testing.T) {
+func TestHCLAttrsCapsuleType_ObjectVars(t *testing.T) {
 	dt := []byte(`
 	variable "foo" {
 		default = "bar"
@@ -714,6 +716,52 @@ func TestHCLAttrsCapsuleTypeVars(t *testing.T) {
 	require.Equal(t, []string{"type=oci,dest=../bar.tar"}, stringify(web.Outputs))
 	require.Equal(t, []string{"type=local,src=path/to/cache", "user/app:cache"}, stringify(web.CacheFrom))
 	require.Equal(t, []string{"id=oci,src=/local/secret"}, stringify(web.Secrets))
+}
+
+func TestHCLAttrsCapsuleType_MissingVars(t *testing.T) {
+	dt := []byte(`
+	target "app" {
+		attest = [
+			"type=sbom,disabled=${SBOM}",
+		]
+
+		cache-from = [
+			{ type = "registry", ref = "user/app:${FOO1}" },
+      "type=local,src=path/to/cache:${FOO2}",
+		]
+
+		cache-to = [
+			{ type = "local", dest = "path/to/${BAR}" },
+		]
+
+		output = [
+			{ type = "oci", dest = "../${OUTPUT}.tar" },
+		]
+
+		secret = [
+			{ id = "mysecret", src = "/local/${SECRET}" },
+		]
+
+		ssh = [
+			{ id = "key", paths = ["path/to/${SSH_KEY}"] },
+		]
+	}
+	`)
+
+	var diags hcl.Diagnostics
+	_, err := ParseFile(dt, "docker-bake.hcl")
+	require.ErrorAs(t, err, &diags)
+
+	re := regexp.MustCompile(`There is no variable named "([\w\d_]+)"`)
+	var actual []string
+	for _, diag := range diags {
+		if m := re.FindStringSubmatch(diag.Error()); m != nil {
+			actual = append(actual, m[1])
+		}
+	}
+	require.ElementsMatch(t,
+		[]string{"SBOM", "FOO1", "FOO2", "BAR", "OUTPUT", "SECRET", "SSH_KEY"},
+		actual)
 }
 
 func TestHCLMultiFileAttrs(t *testing.T) {

--- a/go.mod
+++ b/go.mod
@@ -1,6 +1,6 @@
 module github.com/docker/buildx
 
-go 1.22.0
+go 1.23.0
 
 require (
 	github.com/Masterminds/semver/v3 v3.2.1

--- a/hack/dockerfiles/lint.Dockerfile
+++ b/hack/dockerfiles/lint.Dockerfile
@@ -7,7 +7,7 @@ ARG XX_VERSION=1.6.1
 ARG GOLANGCI_LINT_VERSION=1.62.0
 ARG GOPLS_VERSION=v0.26.0
 # disabled: deprecated unusedvariable simplifyrange
-ARG GOPLS_ANALYZERS="embeddirective fillreturns infertypeargs nonewvars norangeoverfunc noresultvalues simplifycompositelit simplifyslice undeclaredname unusedparams useany"
+ARG GOPLS_ANALYZERS="embeddirective fillreturns infertypeargs nonewvars noresultvalues simplifycompositelit simplifyslice undeclaredname unusedparams useany"
 
 FROM --platform=$BUILDPLATFORM tonistiigi/xx:${XX_VERSION} AS xx
 

--- a/util/buildflags/attests_cty.go
+++ b/util/buildflags/attests_cty.go
@@ -24,9 +24,7 @@ func (e *Attests) FromCtyValue(in cty.Value, p cty.Path) error {
 
 func (e *Attests) fromCtyValue(in cty.Value, p cty.Path) error {
 	*e = make([]*Attest, 0, in.LengthInt())
-	for elem := in.ElementIterator(); elem.Next(); {
-		_, value := elem.Element()
-
+	for value := range eachElement(in) {
 		entry := &Attest{}
 		if err := entry.FromCtyValue(value, p); err != nil {
 			return err
@@ -64,6 +62,10 @@ func (e *Attest) FromCtyValue(in cty.Value, p cty.Path) error {
 	e.Attrs = map[string]string{}
 	for it := conv.ElementIterator(); it.Next(); {
 		k, v := it.Element()
+		if !v.IsKnown() {
+			continue
+		}
+
 		switch key := k.AsString(); key {
 		case "type":
 			e.Type = v.AsString()

--- a/util/buildflags/cache_cty.go
+++ b/util/buildflags/cache_cty.go
@@ -23,13 +23,7 @@ func (o *CacheOptions) FromCtyValue(in cty.Value, p cty.Path) error {
 
 func (o *CacheOptions) fromCtyValue(in cty.Value, p cty.Path) error {
 	*o = make([]*CacheOptionsEntry, 0, in.LengthInt())
-	for elem := in.ElementIterator(); elem.Next(); {
-		_, value := elem.Element()
-
-		if isEmpty(value) {
-			continue
-		}
-
+	for value := range eachElement(in) {
 		// Special handling for a string type to handle ref only format.
 		if value.Type() == cty.String {
 			entries, err := ParseCacheEntry([]string{value.AsString()})

--- a/util/buildflags/export_cty.go
+++ b/util/buildflags/export_cty.go
@@ -23,13 +23,7 @@ func (e *Exports) FromCtyValue(in cty.Value, p cty.Path) error {
 
 func (e *Exports) fromCtyValue(in cty.Value, p cty.Path) error {
 	*e = make([]*ExportEntry, 0, in.LengthInt())
-	for elem := in.ElementIterator(); elem.Next(); {
-		_, value := elem.Element()
-
-		if isEmpty(value) {
-			continue
-		}
-
+	for value := range eachElement(in) {
 		entry := &ExportEntry{}
 		if err := entry.FromCtyValue(value, p); err != nil {
 			return err

--- a/util/buildflags/secrets_cty.go
+++ b/util/buildflags/secrets_cty.go
@@ -30,13 +30,7 @@ func (s *Secrets) FromCtyValue(in cty.Value, p cty.Path) error {
 
 func (s *Secrets) fromCtyValue(in cty.Value, p cty.Path) error {
 	*s = make([]*Secret, 0, in.LengthInt())
-	for elem := in.ElementIterator(); elem.Next(); {
-		_, value := elem.Element()
-
-		if isEmpty(value) {
-			continue
-		}
-
+	for value := range eachElement(in) {
 		entry := &Secret{}
 		if err := entry.FromCtyValue(value, p); err != nil {
 			return err
@@ -71,13 +65,13 @@ func (e *Secret) FromCtyValue(in cty.Value, p cty.Path) error {
 		return err
 	}
 
-	if id := conv.GetAttr("id"); !id.IsNull() {
+	if id := conv.GetAttr("id"); !id.IsNull() && id.IsKnown() {
 		e.ID = id.AsString()
 	}
-	if src := conv.GetAttr("src"); !src.IsNull() {
+	if src := conv.GetAttr("src"); !src.IsNull() && src.IsKnown() {
 		e.FilePath = src.AsString()
 	}
-	if env := conv.GetAttr("env"); !env.IsNull() {
+	if env := conv.GetAttr("env"); !env.IsNull() && env.IsKnown() {
 		e.Env = env.AsString()
 	}
 	return nil

--- a/util/buildflags/ssh_cty.go
+++ b/util/buildflags/ssh_cty.go
@@ -30,13 +30,7 @@ func (s *SSHKeys) FromCtyValue(in cty.Value, p cty.Path) error {
 
 func (s *SSHKeys) fromCtyValue(in cty.Value, p cty.Path) error {
 	*s = make([]*SSH, 0, in.LengthInt())
-	for elem := in.ElementIterator(); elem.Next(); {
-		_, value := elem.Element()
-
-		if isEmpty(value) {
-			continue
-		}
-
+	for value := range eachElement(in) {
 		entry := &SSH{}
 		if err := entry.FromCtyValue(value, p); err != nil {
 			return err
@@ -71,10 +65,10 @@ func (e *SSH) FromCtyValue(in cty.Value, p cty.Path) error {
 		return err
 	}
 
-	if id := conv.GetAttr("id"); !id.IsNull() {
+	if id := conv.GetAttr("id"); !id.IsNull() && id.IsKnown() {
 		e.ID = id.AsString()
 	}
-	if paths := conv.GetAttr("paths"); !paths.IsNull() {
+	if paths := conv.GetAttr("paths"); !paths.IsNull() && paths.IsKnown() {
 		if err := gocty.FromCtyValue(paths, &e.Paths); err != nil {
 			return err
 		}

--- a/util/buildflags/utils.go
+++ b/util/buildflags/utils.go
@@ -1,6 +1,8 @@
 package buildflags
 
 import (
+	"iter"
+
 	"github.com/zclconf/go-cty/cty"
 	"github.com/zclconf/go-cty/cty/gocty"
 )
@@ -34,7 +36,7 @@ func removeDupes[E comparable[E]](s []E) []E {
 }
 
 func getAndDelete(m map[string]cty.Value, attr string, gv interface{}) error {
-	if v, ok := m[attr]; ok {
+	if v, ok := m[attr]; ok && v.IsKnown() {
 		delete(m, attr)
 		return gocty.FromCtyValue(v, gv)
 	}
@@ -44,11 +46,28 @@ func getAndDelete(m map[string]cty.Value, attr string, gv interface{}) error {
 func asMap(m map[string]cty.Value) map[string]string {
 	out := make(map[string]string, len(m))
 	for k, v := range m {
-		out[k] = v.AsString()
+		if v.IsKnown() {
+			out[k] = v.AsString()
+		}
 	}
 	return out
 }
 
-func isEmpty(v cty.Value) bool {
-	return v.Type() == cty.String && v.AsString() == ""
+func isEmptyOrUnknown(v cty.Value) bool {
+	return !v.IsKnown() || (v.Type() == cty.String && v.AsString() == "")
+}
+
+func eachElement(in cty.Value) iter.Seq[cty.Value] {
+	return func(yield func(v cty.Value) bool) {
+		for elem := in.ElementIterator(); elem.Next(); {
+			_, value := elem.Element()
+			if isEmptyOrUnknown(value) {
+				continue
+			}
+
+			if !yield(value) {
+				return
+			}
+		}
+	}
 }

